File: 54-spiral-matrix/54-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        
        vector<int> ans ;
        
        int m = matrix.size();
        int n = matrix[0].size() ; 
        int row = m-1 ; 
        int col = n-1 ;
        int direction = 1;
     
        int r = 0 , c = 0 ;
        while ( r <= row  && c <= col) {
            if(direction == 1){
                for(int  j = c ; j <= col ; j++) 
                    ans.push_back(matrix[r][j]);
                r++; 
                direction = 2;
            }
             
            else if(direction == 2){
                for(int i = r; i <= row ; i++) 
                    ans.push_back(matrix[i][col]);
                col--;
                direction = 3;
            }
            
            else if(direction == 3){
               for(int j = col ; j >= c  ; j--) 
                   ans.push_back(matrix[row][j]) ; 
                   row--; 
                direction = 4;
               
            }
            else if(direction == 4){
                for( int i = row ; i >= r ; i--) 
                    ans.push_back(matrix[i][c]);
                c++;
                direction = 1;
            }
            
               
        }
        return ans ;
    }
};

File: 54-spiral-matrix/54-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    
    let ans = [];
    
    let m = matrix.length;
    let n = matrix[0].length;
    
    let top = 0 , down = m-1;
    let left = 0 , right = n-1;
    let dir = 0;
    
    while(top <= down && left <= right){
        
        if(dir == 0){
            for(let i = left ; i <= right ; i++ ){
                ans.push(matrix[top][i]);
            }
            top = top+1;
        }
        else if(dir == 1) {
            for(let i = top ; i <= down ; i++){
                ans.push(matrix[i][right]);
            }
            right = right-1;
        }
        else if(dir == 2){
            for(let i = right ; i >= left ; i--){
                ans.push(matrix[down][i]);
            }
            down = down-1;
        }
        else if(dir == 3){
            for(let i = down ; i >= top ; i--){
                ans.push(matrix[i][left]);
            }
            left = left+1;
        }
        dir = (dir+1)%4;
    }
    return ans;
    
};